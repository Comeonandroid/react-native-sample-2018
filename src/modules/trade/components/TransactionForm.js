import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { ifProp, prop } from 'styled-tools'

import Yup from '../../../utils/yup'
import Icon from '../../../utils/icon'
import { t } from '../../../locales/i18n'

import { View, Text } from '../../../elements'
import { BaseForm, FormContainer, FormGroup, SubmitContainer, ValidationError } from '../../../elements'
import { SectionHeader } from '../../../elements'
import { UnitsTextfield } from '../../../elements'
import { SuccessButton, DangerButton } from '../../../elements'

import FeeSelect from './FeeSelect'

const textInputStyle = {
  fontStyle: 'italic'
}

export default class TransactionForm extends PureComponent {

  constructor(props) {
    super(props)

    this.validationSchema = this.validationSchema.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
    this.renderForm = this.renderForm.bind(this)
  }

  validationSchema() {
    return Yup.object().shape({})
  }

  handleSubmit(values, actions) {
    this.props.onSend(values)
    actions.setSubmitting(false)
  }

  renderForm(props) {
    const { type } = this.props
    const {
      values,
      errors,
      touched,
      setFieldValue,
      isSubmitting,
      setFieldTouched,
      handleSubmit
    } = props
    return (
      <Wrapper>
        <Cell>
          <TransactionFormContainer>
            <FormHeaderContainer>
              <Icon size={16} name='wallet' />
              <TransactionSectionLabel right>2.214522 WETH</TransactionSectionLabel>
            </FormHeaderContainer>
            <FormGroup top={9} bottom={16}>
              <TransactionSectionHeader>{t('wallet.price')}</TransactionSectionHeader>
              <UnitsTextfield
                type='number'
                name='price'
                value={values.price}
                error={errors.price && touched.price}
                placeholder={t('inputs.amount.placeholder')}
                units='WETH'
                size='small'
                textInputStyle={textInputStyle}
                onBlur={setFieldTouched}
                onChange={setFieldValue}>
              </UnitsTextfield>
              {errors.price && touched.price && <ValidationError>{errors.price}</ValidationError>}
            </FormGroup>
            <FormGroup top={0}>
              <TransactionSectionHeader>{t('wallet.amount')}</TransactionSectionHeader>
              <FormHeaderContainer top={0} bottom={7}>
                <MaxLabel>{t('wallet.max')}</MaxLabel>
                <TransactionSectionLabel right>141.214 ZRX</TransactionSectionLabel>
              </FormHeaderContainer>
              <UnitsTextfield
                type='number'
                name='amount'
                value={values.amount}
                error={errors.amount && touched.amount}
                placeholder={t('inputs.amount.placeholder')}
                units='ZRX'
                size='small'
                textInputStyle={textInputStyle}
                onBlur={setFieldTouched}
                onChange={setFieldValue}>
              </UnitsTextfield>
              {errors.amount && touched.amount && <ValidationError>{errors.amount}</ValidationError>}
            </FormGroup>

            <FormHeaderContainer top={2}>
              <TransactionSectionLabel>{t('wallet.all')}</TransactionSectionLabel>
              <TransactionSectionLabel right>1.66666 WETH</TransactionSectionLabel>
            </FormHeaderContainer>

            <FormHeaderContainer top={1}>
              <TransactionSectionSubLabel></TransactionSectionSubLabel>
              <TransactionSectionSubLabel right>7234.00 CNY</TransactionSectionSubLabel>
            </FormHeaderContainer>

            <FormHeaderContainer top={8}>
              <TransactionSectionLabel>{t('wallet.fee')}</TransactionSectionLabel>
              <TransactionSectionLabel right>0.00 ZRX</TransactionSectionLabel>
            </FormHeaderContainer>

            <FormHeaderContainer top={3}>
              <TransactionSectionSubLabel></TransactionSectionSubLabel>
              <TransactionSectionSubLabel right>0.00 CNY</TransactionSectionSubLabel>
            </FormHeaderContainer>
          </TransactionFormContainer>
          <TransactionFormContainer>
            <FormGroup top={6}>
              <TransactionSectionHeader>{t('wallet.gas_price')}</TransactionSectionHeader>
              <TransactionSectionHeader right>0.00000 ETH</TransactionSectionHeader>
            </FormGroup>
          </TransactionFormContainer>
          <FormGroup top={2} bottom={6} >
            <FeeSelect />
          </FormGroup>
        </Cell>
        <Cell>
          <TransactionSubmitContainer>
            {type === 'buy' && <SuccessButton enabled={!isSubmitting} onPress={handleSubmit}>{t('inputs.buy')}</SuccessButton>}
            {type === 'sell' && <DangerButton enabled={!isSubmitting} onPress={handleSubmit}>{t('inputs.sell')}</DangerButton>}
          </TransactionSubmitContainer>
        </Cell>
      </Wrapper>
    )
  }

  render() {
    return (
      <BaseForm
        enableReinitialize
        initialValues={this.props.values}
        validationSchema={this.validationSchema}
        onSubmit={this.handleSubmit}
        render={this.renderForm}
      />
    )
  }
}

TransactionForm.propTypes = {
  onSend: PropTypes.func.isRequired,
  type: PropTypes.string.isRequired,
  values: PropTypes.shape({
    gasLimit: PropTypes.string,
    gasPrice: PropTypes.string,
    price: PropTypes.string,
    amount: PropTypes.string
  })
}

TransactionForm.defaultProps = {
  type: 'buy',
  values: {
    gasLimit: '',
    gasPrice: '',
    price: '',
    amount: '',
  }
}

const Wrapper = styled(View) `
  background: ${prop('theme.colors.background')};
  flex: 1;
  flex-direction: column;
  justify-content: space-between;
`

const Cell = styled(View) `;
`

const TransactionFormContainer = styled(FormContainer) `
  padding: 6px 12px 0px;
  flex-direction: column;
  justify-content: flex-start;
`

const TransactionSubmitContainer = styled(SubmitContainer) `
  margin-top: 0px;
`

const FormHeaderContainer = styled(FormGroup) `
  flex-direction: row;
  align-items: baseline;
  justify-content: space-between;
`

const MaxLabel = styled(Text) `
  font-family: ${prop('theme.typo.condensedFont')};
  color: ${prop('theme.colors.success')};
`

const TransactionSectionHeader = styled(SectionHeader) `
  margin-top: 0px;
  font-family: ${prop('theme.typo.condensedFont')};
  font-size: 15px;
  text-align: ${ifProp('right', 'right', 'left')}
`

const TransactionSectionLabel = styled(TransactionSectionHeader) `
  margin: 0;
  padding: 0;
`

const TransactionSectionSubLabel = styled(TransactionSectionHeader) `
  margin: 0;
  padding: 0;
  font-size: 11px;
`
