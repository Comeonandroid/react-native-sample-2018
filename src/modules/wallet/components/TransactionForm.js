import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { ifProp, prop } from 'styled-tools'

import Yup from '../../../utils/yup'
import { t } from '../../../locales/i18n'

import { View } from '../../../elements'
import { BaseForm, FormContainer, FormGroup, StickyButtonContainer } from '../../../elements'
import { SectionHeader } from '../../../elements'
import { UnitsTextfield } from '../../../elements'
import { SuccessButton, DangerButton } from '../../../elements'

export default class TransactionForm extends PureComponent {

  constructor(props) {
    super(props)

    this.validationSchema = this.validationSchema.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
    this.handleReset = this.handleReset.bind(this)
    this.renderForm = this.renderForm.bind(this)
  }

  validationSchema() {
    return Yup.object().shape({
      gasPrice: Yup.string().required(t('inputs.errors.required')),
      gasLimit: Yup.string().required(t('inputs.errors.required'))
    })
  }

  handleSubmit(values, actions) {
    this.props.onSend(values)
    actions.setSubmitting(false)
  }

  handleReset() {
    this.props.onCancel()
  }

  renderForm(props) {
    const {
      values,
      errors,
      touched,
      setFieldValue,
      isSubmitting,
      handleSubmit,
      handleReset,
      setFieldTouched
    } = props
    return (
      <Wrapper>
        <TransactionFormContainer>
          <TransactionFormGroup>
            <TransactionSectionHeader>{t('wallet.gas_limit')}</TransactionSectionHeader>
            <UnitsTextfield
              type='number'
              name='gasPrice'
              value={values.gasPrice}
              error={errors.gasPrice && touched.gasPrice}
              placeholder={t('inputs.amount.placeholder')}
              units='UNITS'
              size='small'
              onBlur={setFieldTouched}
              onChange={setFieldValue}>
            </UnitsTextfield>
          </TransactionFormGroup>
          <LastTransactionFormGroup>
            <TransactionSectionHeader>{t('wallet.gas_price')}</TransactionSectionHeader>
            <UnitsTextfield
              type='number'
              name='gasLimit'
              value={values.gasLimit}
              error={errors.gasLimit && touched.gasLimit}
              placeholder={t('inputs.amount.placeholder')}
              units='GWEI'
              size='small'
              onBlur={setFieldTouched}
              onChange={setFieldValue}>
            </UnitsTextfield>
          </LastTransactionFormGroup>
          <TransactionLabelContainer>
            <TransactionSectionHeader>{t('wallet.transaction_fee')}</TransactionSectionHeader>
            <TransactionSectionHeader right>0.00000 ETH</TransactionSectionHeader>
          </TransactionLabelContainer>
        </TransactionFormContainer>

        <StickyButtonContainer>
          <SuccessButton enabled={!isSubmitting} onPress={handleSubmit}>{t('inputs.ok')}</SuccessButton>
          <DangerButton onPress={handleReset}>{t('inputs.cancel')}</DangerButton>
        </StickyButtonContainer>
      </Wrapper>
    )
  }

  render() {
    return (
      <BaseForm
        initialValues={this.props.values}
        validationSchema={this.validationSchema}
        onSubmit={this.handleSubmit}
        onReset={this.handleReset}
        render={this.renderForm}
      />
    )
  }
}

TransactionForm.propTypes = {
  onSend: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  values: PropTypes.shape({
    gasLimit: PropTypes.string,
    gasPrice: PropTypes.string
  })
}

TransactionForm.defaultProps = {
  values: {
    gasLimit: '12',
    gasPrice: '10'
  }
}

const Wrapper = styled(View) `
  background: ${prop('theme.colors.background')};
  flex-direction: row;
  justify-content: center;
`

const TransactionFormContainer = styled(FormContainer) `
  width: 56%
`

const TransactionSectionHeader = styled(SectionHeader) `
  margin-top: 0px;
  font-family: ${prop('theme.typo.condensedFont')};
  font-size: 14px;
  text-align: ${ifProp('right', 'right', 'left')}
`

const TransactionFormGroup = styled(FormGroup) `
  margin-top: 10px;
`

const LastTransactionFormGroup = styled(FormGroup) `
  margin-top: 0px;
`

const TransactionLabelContainer = styled(FormGroup) `
  margin-top: 4px;
  margin-bottom: 56px;
`
