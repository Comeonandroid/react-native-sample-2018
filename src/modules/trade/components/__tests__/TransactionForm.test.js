import 'react-native'
import React from 'react'
import { shallow } from 'enzyme'

import { itBehavesLike } from '../../../../specs/shared_examples/index'

import TransactionForm from '../TransactionForm'

const setup = propOverrides => {
  const props = Object.assign({
    type: 'buy',
    onSend: jest.fn(),
    values: {
      gasLimit: '',
      gasPrice: '',
      price: '',
      amount: '',
    }
  }, propOverrides)

  const component = <TransactionForm {...props} />
  const wrapper = shallow(<TransactionForm {...props} />)

  return {
    props,
    wrapper,
    component
  }
}

describe('<TransactionForm />', () => {
  it('renders correctly', () => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
  })

  // describe('values', () => {
  //   const { component } = setup({
  //     values: {
  //       gasLimit: '12',
  //       gasPrice: '10',
  //       price: '5',
  //       amount: '9',
  //     }
  //   })
  //   itBehavesLike('aThemedComponent', { subject: component })
  // })

  // describe('type: sell', () => {
  //   const { component } = setup({ type: 'sell' })
  //   itBehavesLike('aThemedComponent', { subject: component })
  // })
})
