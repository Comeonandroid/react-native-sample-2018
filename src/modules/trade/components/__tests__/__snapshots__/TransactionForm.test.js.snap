// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TransactionForm /> renders correctly 1`] = `
<BaseForm
  enableReinitialize={true}
  initialValues={
    Object {
      "amount": "",
      "gasLimit": "",
      "gasPrice": "",
      "price": "",
    }
  }
  onSubmit={[Function]}
  render={[Function]}
  validationSchema={[Function]}
/>
`;
